src/v/storage/offset_to_filepos_consumer.h

```diff
@@ -26,69 +26,41 @@ namespace storage::internal {
-            return ss::make_ready_future<ss::stop_iteration>(
-              ss::stop_iteration::yes);
-        }
-        // if we are in the middle of an actual batch, throw an exception to
-        // preserve previous semantics
 
+        if (_target_last_offset <= batch.base_offset()) {
+            _filepos = {_prev_batch_last_offset, _prev};
+            co_return ss::stop_iteration::yes;
+        }
         if (
           _target_last_offset > batch.base_offset()
-          && _target_last_offset < batch.last_offset()) {
-            _filepos = std::nullopt;
-            return ss::make_exception_future<ss::stop_iteration>(
-              std::runtime_error{fmt::format(
-                "offset_to_filepos_consumer can only translate base_offsets. "
-                "Current batch's {}-{}. Requested offset {} is in a middle of "
-                "curren batch.",
-                batch.base_offset(),
-                batch.last_offset(),
-                _target_last_offset)});
-        }
-        if (_target_last_offset == batch.last_offset()) {
-            return ss::make_exception_future<ss::stop_iteration>(
-              std::runtime_error{fmt::format(
-                "offset_to_filepos_consumer can only translate base_offsets. "
-                "Current batch's {}-{}. last_offset matches target offset:{}. "
-                "accumulated bytes:{}",
-                batch.base_offset(),
-                batch.last_offset(),
-                _target_last_offset,
-                _accumulator)});
+          && _target_last_offset <= batch.last_offset()) {
+            throw std::runtime_error(fmt::format(
+              "Offset to file position consumer isn't able to translate "
+              "offsets other than batch base offset or offset being in the "
+              "gap. Requested offset: {}, current batch offsets: [{},{}]",
+              _target_last_offset,
+              batch.base_offset(),
+              batch.last_offset()));
         }
-        if (batch.base_offset() > _target_last_offset) {
-            return ss::make_exception_future<ss::stop_iteration>(
-              std::runtime_error{fmt::format(
-                "offset_to_filepos_consumer can only translate base_offsets. "
-                "Current batch's {}-{}, exceeds target offset:{}. accumulated "
-                "bytes:{}",
-                batch.base_offset(),
-                batch.last_offset(),
-                _target_last_offset,
-                _accumulator)});
-        }
-        /**
-         * We update _filepos here in case we hit a gap, if so we need to return
-         * an offset and end position of a last batch preceding the gap.
-         */
-        _filepos = {batch.last_offset(), _accumulator};
-        return ss::make_ready_future<ss::stop_iteration>(
-          ss::stop_iteration::no);
+
+        _prev_batch_last_offset = batch.last_offset();
+        co_return ss::stop_iteration::no;
```

Comment:
hmm, doesn't _filepos always need to be updated upon return so that the caller makes some sort of progress?